### PR TITLE
[FOLLOWUP] eliminate the duplicated sort keys in Order By clause

### PR DESCRIPTION
**Pull request by mingmwang**
# Which issue does this PR close?

Closes #.
Follow up mentioned in the PR
https://github.com/apache/arrow-datafusion/pull/5462

# Rationale for this change

Dedup sort keys should ignoring sort options:

For example `select * from t1 order by id desc, id, name, id asc`;
The real sort key should be: `Sort Key: id DESC, name`

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by mingmwang** [comment]
@jackwener

File: datafusion/optimizer/src/eliminate_duplicated_expr.rs [hunk 1]
```diff
@@ -100,4 +115,23 @@ mod tests {
...
     }
+
+    #[test]
+    fn eliminate_sort_exprs_with_options() -> Result<()> {
+        let table_scan = test_table_scan().unwrap();
+        let sort_exprs = vec![
+            Expr::Sort(ExprSort::new(Box::new(col("a")), true, true)),
+            Expr::Sort(ExprSort::new(Box::new(col("b")), true, false)),
+            Expr::Sort(ExprSort::new(Box::new(col("a")), false, false)),
+            Expr::Sort(ExprSort::new(Box::new(col("b")), false, true)),
```
Comment (jackwener) [inline datafusion/optimizer/src/eliminate_duplicated_expr.rs:126]:
```suggestion
            col("a").sort(true, true),
            col("b").sort(true, false),
            col("a").sort(false, false),
            col("b").sort(false, true),
```

**Review by jackwener — APPROVED** [review]
Great job👍

File: datafusion/optimizer/src/eliminate_duplicated_expr.rs [hunk 2]
```diff
@@ -18,8 +18,9 @@
 use crate::optimizer::ApplyOrder;
 use crate::{OptimizerConfig, OptimizerRule};
 use datafusion_common::Result;
+use datafusion_expr::expr::Sort as ExprSort;
```
Comment (jackwener) [inline datafusion/optimizer/src/eliminate_duplicated_expr.rs:21]:
```suggestion
```

File: datafusion/optimizer/src/eliminate_duplicated_expr.rs [hunk 3]
```diff
@@ -41,15 +42,28 @@ impl OptimizerRule for EliminateDuplicatedExpr {
     ) -> Result<Option<LogicalPlan>> {
         match plan {
             LogicalPlan::Sort(sort) => {
+                let normalized_sort_keys = sort
+                    .expr
+                    .iter()
+                    .map(|e| match e {
+                        Expr::Sort(ExprSort { expr, .. }) => {
+                            Expr::Sort(ExprSort::new(expr.clone(), true, false))
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_duplicated_expr.rs:50]: I am not sure about always using `true`, `fase`

Using the example
```sql
select * from t1 order by id desc, id, name, id asc;
select * from t1 order by id asc, id, name, id desc;
```

Is the sort key always the same ? It seems like in the first query the sort key should be `id desc, name` and the second it should be `id asc, name`

Maybe I am missing something

Comment (mingmwang) [reply datafusion/optimizer/src/eliminate_duplicated_expr.rs:50]: Your conclusion is right.  The `true`, `fase` is just for normalize and dedup.

Comment (alamb) [reply datafusion/optimizer/src/eliminate_duplicated_expr.rs:50]: Maybe we can add a test to cover this case as well?

File: datafusion/optimizer/src/eliminate_duplicated_expr.rs [hunk 4]
```diff
@@ -41,15 +42,28 @@ impl OptimizerRule for EliminateDuplicatedExpr {
...
                 let mut dedup_expr = Vec::new();
                 let mut dedup_set = HashSet::new();
-                for expr in &sort.expr {
-                    if !dedup_set.contains(expr) {
-                        dedup_expr.push(expr);
-                        dedup_set.insert(expr.clone());
-                    }
-                }
+                sort.expr.iter().zip(normalized_sort_keys.iter()).for_each(
+                    |(expr, normalized_expr)| {
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_duplicated_expr.rs:60]: I think I missed the case that the expr/normalized expr match 👍

**Review by alamb — APPROVED** [review]
I think the one test showing some other `NULLs ASC` would be valuable for this PR, but the code looks good. Thank you for this @mingmwang and @jackwener

File: datafusion/optimizer/src/eliminate_duplicated_expr.rs (hunk 3 above)
Comment (mingmwang) [reply datafusion/optimizer/src/eliminate_duplicated_expr.rs:50]: > Maybe we can add a test to cover this case as well?

Done.

**Review by alamb — APPROVED** [review]
Thanks @mingmwang

**Pull request merged**